core/model/impl/src/main/kotlin/parsing/BooleanExpressionParser.kt

```diff
@@ -0,0 +1,302 @@
+ * limitations under the License.
+ */
+
+package com.yandex.yatagan.core.model.impl.parsing
+
+import com.yandex.yatagan.base.enumMapOf
+import com.yandex.yatagan.base.enumSetOf
+import com.yandex.yatagan.base.forEachBi
+import com.yandex.yatagan.core.model.impl.parsing.BooleanExpressionParser.Token.Kind.BeginOfExpression
+import com.yandex.yatagan.core.model.impl.parsing.BooleanExpressionParser.Token.Kind.BinaryAnd
+import com.yandex.yatagan.core.model.impl.parsing.BooleanExpressionParser.Token.Kind.BinaryOr
+import com.yandex.yatagan.core.model.impl.parsing.BooleanExpressionParser.Token.Kind.EndOfExpression
+import com.yandex.yatagan.core.model.impl.parsing.BooleanExpressionParser.Token.Kind.LeftParenthesis
+import com.yandex.yatagan.core.model.impl.parsing.BooleanExpressionParser.Token.Kind.RightParenthesis
+import com.yandex.yatagan.core.model.impl.parsing.BooleanExpressionParser.Token.Kind.UnaryNot
+import com.yandex.yatagan.core.model.impl.parsing.BooleanExpressionParser.Token.Kind.Variable
+
+internal class BooleanExpressionParser<E>(
+    private val expressionSource: String,
+    private val factory: Factory<E>,
+) {
+    interface Factory<E> {
+        sealed interface ParseResult<out E> {
+            class Ok<E>(val variable: E) : ParseResult<E>
+            class Error(val message: CharSequence) : ParseResult<Nothing>
+        }
+
+        fun createAnd(lhs: E, rhs: E): E
+        fun createOr(lhs: E, rhs: E): E
+        fun createNot(e: E): E
+        fun parseVariable(text: String): ParseResult<E>
+    }
+
+    @Throws(ParseException::class)
+    fun parse(): E {
+        // 1. Lex the tokens
+        val tokens = lex()
+
+        // 2. Check the syntax, throw if invalid
+        validate(tokens)
+
+        // 3. Convert to inverse postfix, throw if semantic errors detected
+        val inversePostfix = toInversePostfix(tokens)
+
+        // 4. Build AST
+        return buildExpressionTree(inversePostfix)
+    }
+
+    private fun lex(): List<Token> {
+        val tokens = arrayListOf(
+            Token(
+                kind = BeginOfExpression,
+                sourceLocation = 0,
+                text = "",
+            )
+        )
+        var variableIndex = -1
+        val variableBuilder = StringBuilder()
+
+        fun commitVariableTokenIfAny() {
+            if (variableIndex >= 0) {
+                tokens += Token(
+                    kind = Variable,
+                    sourceLocation = variableIndex,
+                    text = variableBuilder.toString(),
+                )
+                variableBuilder.clear()
+                variableIndex = -1
+            }
+        }
+
+        expressionSource.forEachIndexed { index, char ->
+            when (char) {
+                '(' -> tokens += Token(
+                    kind = LeftParenthesis,
+                    sourceLocation = index,
+                    text = "(",
+                ).also { commitVariableTokenIfAny() }
```

Comment:
can variables be before opening brackets?

Reply:
Yes, syntax checking is done later, lexer only checks on the character level, no on the token level.